### PR TITLE
add i18n extraction script and document the use

### LINKED ISSUE #1189: Can't extract msgids from mockup to create a PO file
In previous versions of mockup there was a `make i18n-dump` script that called a corresponding task in `Gruntfile.js` to create a widgets.pot file that could be copied to `plone.app.locales` and then resync the translation files.

See:
- https://github.com/plone/mockup/blob/4.x/Makefile#L106-L107
- https://github.com/plone/mockup/blob/4.x/mockup/Gruntfile.js#L117

There is nothing like that now in mockup.

We need this to update the translations for the upcoming Plone 6 version.

Related issues:
- https://github.com/collective/plone.app.locales/issues/358
- https://github.com/plone/mockup/issues/1147

### PR DESCRIPTION
**Pull request by erral**
Fixes #1189 

I am not at all a JS guy, I just copied the script that used to work in Mockup 4 and it seems to update the widgets.pot file.

Can someone review this please?

@thet @petschki @mauritsvanrees

### CODE REVIEW

File: src/i18n.js [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+            var stats = fs.statSync(path + filename);
+            if (stats.isDirectory()) {
+                checkDir(path + filename + "/");
+            } else {
+                checkFile(path + filename);
+            }
+        });
+    };
+    checkDir("./");
+    fs.writeFileSync("../widgets.pot", output);
```
Comment (mauritsvanrees) [inline src/i18n.js:75]: This needs to be `widgets.pot` without `../`.
At least, on my system, the pot file ends up outside of the repository, in the directory just above it. It does not matter whether I am in `mockup` or in `mockup/src`. I guess npm changes to the directory where the `package.json` lives`.

File: README.md [hunk 2]
```diff
@@ -180,3 +180,28 @@ https://www.npmjs.com/package/webpack-bundle-analyzer
...
+
+```
+npm run i18n
+```
+
+2. Copy the widgets.pot file to plone.app.locales
+
+Assuming you are doing this from buildout.coredev:
+```
+cp widgets.pot ../src/plone.app.locales/plone/app/locales/locales/
```
Comment (mauritsvanrees) [inline README.md:199]: I expect the current directory to be `coredev-checkout/src/mockup`. Then the command should be this:

```
cp widgets.pot ../plone.app.locales/plone/app/locales/locales/
```

**Review by mauritsvanrees — CHANGES_REQUESTED** [review]
Seems to work, except for two paths that should be changed, see inline comments.
And the `widgets.pot` file that you have added, is still the old one.

**Comment by mauritsvanrees** [comment]
@thet, @petschki Translations would be missing for the following patterns, which should be old ones: `filemanager`, `pickadate`, `resourceregistry`, `thememapper`, `toolbar`. Does that look okay?

Is anything maybe moved to `patternslib` and do we need to move some translations there, or include patternslib in the files where we extract from?

Some statistics.
With the old file, these are the patterns and the number of messages:

```
$ grep /patterns/ widgets.pot | cut -d "/" -f-3 | sort | uniq -c
  31 #: ./patterns/filemanager
   9 #: ./patterns/livesearch
   2 #: ./patterns/modal
   9 #: ./patterns/pickadate
   8 #: ./patterns/querystring
   9 #: ./patterns/relateditems
  83 #: ./patterns/resourceregistry
  47 #: ./patterns/structure
  22 #: ./patterns/thememapper
  31 #: ./patterns/tinymce
   3 #: ./patterns/toolbar
   9 #: ./patterns/upload
$ grep -c msgid widgets.pot 
264
```

and here for the new file:

```
$ grep /src/pat widgets.pot | cut -d "/" -f-4 | sort | uniq -c
   9 #: ./src/pat/livesearch
   3 #: ./src/pat/manageportlets
   2 #: ./src/pat/modal
   8 #: ./src/pat/querystring
  10 #: ./src/pat/relateditems
  50 #: ./src/pat/structure
  28 #: ./src/pat/tinymce
   9 #: ./src/pat/upload
$ grep -c msgid widgets.pot 
120
```

**Comment by erral** [comment]
Updated stats:

```
$ grep /src/pat widgets.pot | cut -d "/" -f-4 | sort | uniq -c
      1 #: ./src/pat/formunloadalert
      9 #: ./src/pat/livesearch
      3 #: ./src/pat/manageportlets
      2 #: ./src/pat/modal
      1 #: ./src/pat/preventdoublesubmit
      8 #: ./src/pat/querystring
     10 #: ./src/pat/relateditems
     50 #: ./src/pat/structure
     30 #: ./src/pat/tinymce
      9 #: ./src/pat/upload
$ grep -c msgid widgets.pot 
124
```

**Review by petschki — CHANGES_REQUESTED** [review]
Very cool!

I suggest to use `yarn i18n` in the documentation (istead of npm run ...) since we always use `yarn` for the other commands.

The single line `_t()` probably gets refactored by the eslinter ... if this is mandatory for i18n we have to tell him ... @thet can this be configured?

**Comment by thet** [comment]
super, big +1 for me. I hadn't this on my radar.
I'm currently fixing some stuff, so I put this to draft to finish my changes.

**Comment by thet** [comment]
that looks ok. filemanager, pickadate, resourceregistry and thememapper are gone, toolbar is now mostly rendered via page templates.

> @thet, @petschki Translations would be missing for the following patterns, which should be old ones: `filemanager`, `pickadate`, `resourceregistry`, `thememapper`, `toolbar`. Does that look okay?

**Comment by thet** [comment]
@erral thanks a lot again for this contribution. works great!
I‌ did some changes and amended it to your commits.
basically the starting directory is "./src/" - so node_modules, docs, .git, dist etc are not included.
also, i extended the README with the info from @vincentfretin which I‌ accidentally removed some time ago. 

>‌I suggest to use yarn i18n in the documentation (istead of npm run ...) since we always use yarn for the other commands.
done

>‌ The single line _t() probably gets refactored by the eslinter ... if this is mandatory for i18n we have to tell him ... @thet can this be configured?
I‌ added a `` // prettier-ignore`` at the end of those statements. Also, I did search for other statemetns too (``nvim $(rg "_t\(")``) and fixed them.
A better solution would be to use a regex which extracts the text between ``_t()`` over multiple lines, but we can check if we want that optimization sometime when it's necessary.

**Comment by wesleybl** [comment]
> yes, this is still necessary.

It might be interesting to put this in the Readme.

**Pull request merged**